### PR TITLE
add single table conflict check option

### LINKED ISSUE #11837: Remove read data source from readwrite-split zk config throw exception
### Which version of ShardingSphere did you use?

master

### Which project did you use? ShardingSphere-JDBC or ShardingSphere-Proxy?

ShardingSphere-Proxy

### Expected behavior

Update config successful

### Actual behavior

original config in zk

```
- !READWRITE_SPLITTING
  dataSources:
    pr_ds:
      readDataSourceNames:
      - read_ds_0
      - read_ds_1
      writeDataSourceName: write_ds
```

changed config in zk

```
- !READWRITE_SPLITTING
  dataSources:
    pr_ds:
      readDataSourceNames:
      - read_ds_0
      writeDataSourceName: write_ds
```

exception thrown

```log
严重: Exception thrown by subscriber method renew(org.apache.shardingsphere.governance.core.registry.config.event.rule.RuleConfigurationsChangedEvent) on subscriber org.apache.shardingsphere.governance.context.ClusterContextManager@21d1b321 when dispatching event: org.apache.shardingsphere.governance.core.registry.config.event.rule.RuleConfigurationsChangedEvent@3bdfdb40
java.lang.IllegalStateException: Single table conflict, there are multiple tables `t_order` existed.
	at com.google.common.base.Preconditions.checkState(Preconditions.java:589)
	at org.apache.shardingsphere.infra.rule.single.SingleTableDataNodeLoader.lambda$load$0(SingleTableDataNodeLoader.java:53)
	at java.util.HashMap$KeySet.forEach(HashMap.java:933)
	at org.apache.shardingsphere.infra.rule.single.SingleTableDataNodeLoader.load(SingleTableDataNodeLoader.java:53)
	at org.apache.shardingsphere.infra.rule.single.SingleTableRule.<init>(SingleTableRule.java:55)
	at org.apache.shardingsphere.infra.rule.builder.single.SingleTableRuleBuilder.build(SingleTableRuleBuilder.java:40)
	at org.apache.shardingsphere.infra.rule.builder.single.SingleTableRuleBuilder.build(SingleTableRuleBuilder.java:35)
	at org.apache.shardingsphere.infra.rule.builder.ShardingSphereRulesBuilder.buildSchemaRules(ShardingSphereRulesBuilder.java:72)
	at org.apache.shardingsphere.infra.context.metadata.MetaDataContextsBuilder.build(MetaDataContextsBuilder.java:98)
	at org.apache.shardingsphere.governance.context.ClusterContextManager.getChangedMetaData(ClusterContextManager.java:329)
	at org.apache.shardingsphere.governance.context.ClusterContextManager.renew(ClusterContextManager.java:223)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at com.google.common.eventbus.Subscriber.invokeSubscriberMethod(Subscriber.java:87)
	at com.google.common.eventbus.Subscriber$SynchronizedSubscriber.invokeSubscriberMethod(Subscriber.java:144)
	at com.google.common.eventbus.Subscriber$1.run(Subscriber.java:72)
	at com.google.common.util.concurrent.DirectExecutor.execute(DirectExecutor.java:30)
	at com.google.common.eventbus.Subscriber.dispatchEvent(Subscriber.java:67)
	at com.google.common.eventbus.Dispatcher$PerThreadQueuedDispatcher.dispatch(Dispatcher.java:108)
	at com.google.common.eventbus.EventBus.post(EventBus.java:212)
	at java.util.Optional.ifPresent(Optional.java:159)
	at org.apache.shardingsphere.governance.core.registry.GovernanceWatcherFactory.lambda$watch$0(GovernanceWatcherFactory.java:59)
	at org.apache.shardingsphere.governance.repository.zookeeper.CuratorZookeeperRepository.lambda$watch$1(CuratorZookeeperRepository.java:247)
	at org.apache.curator.framework.recipes.cache.CuratorCacheImpl.lambda$putStorage$6(CuratorCacheImpl.java:274)
	at org.apache.curator.framework.listen.MappingListenerManager.lambda$forEach$0(MappingListenerManager.java:92)
	at org.apache.curator.framework.listen.MappingListenerManager.forEach(MappingListenerManager.java:89)
	at org.apache.curator.framework.listen.StandardListenerManager.forEach(StandardListenerManager.java:89)
	at org.apache.curator.framework.recipes.cache.CuratorCacheImpl.lambda$callListeners$10(CuratorCacheImpl.java:293)
	at java.util.concurrent.CompletableFuture$AsyncRun.run(CompletableFuture.java:1640)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
```

### PR DESCRIPTION
**Pull request by strongduanmu**
Fixes #11837.

Changes proposed in this pull request:
- add single table conflict check option
- add test case for single table loader with check option
- add `ShardingSphereRulesBuilderMaterials` to reduced parameters

### CODE REVIEW

File: shardingsphere-infra/shardingsphere-infra-context/src/main/java/org/apache/shardingsphere/infra/context/metadata/MetaDataContextsBuilder.java [hunk 1]
```diff
@@ -37,6 +37,7 @@
 import org.apache.shardingsphere.infra.optimize.context.OptimizeContextFactory;
 import org.apache.shardingsphere.infra.rule.ShardingSphereRule;
 import org.apache.shardingsphere.infra.rule.builder.ShardingSphereRulesBuilder;
+import org.apache.shardingsphere.infra.rule.builder.ShardingSphereRulesBuilderMaterials;
```
Comment (tristaZero) [inline shardingsphere-infra/shardingsphere-infra-context/src/main/java/org/apache/shardingsphere/infra/context/metadata/MetaDataContextsBuilder.java:40]: materials? or material?

File: shardingsphere-infra/shardingsphere-infra-common/src/main/java/org/apache/shardingsphere/infra/config/properties/ConfigurationPropertyKey.java [hunk 2]
```diff
@@ -132,7 +132,12 @@
      * Proxy backend query fetch size. A larger value may increase the memory usage of ShardingSphere Proxy.
      * The default value is -1, which means set the minimum value for different JDBC drivers.
      */
-    PROXY_BACKEND_QUERY_FETCH_SIZE("proxy-backend-query-fetch-size", "-1", int.class);
+    PROXY_BACKEND_QUERY_FETCH_SIZE("proxy-backend-query-fetch-size", "-1", int.class),
+    
+    /**
+     * Whether check single table.
+     */
+    CHECK_SINGLE_TABLE_ENABLED("check-single-table-enabled", String.valueOf(Boolean.FALSE), boolean.class);
```
Comment (tristaZero) [inline shardingsphere-infra/shardingsphere-infra-common/src/main/java/org/apache/shardingsphere/infra/config/properties/ConfigurationPropertyKey.java:140]: how about 'allow tables with the same name exist '? Or something like that.

**Review by tristaZero — CHANGES_REQUESTED** [review]

File: shardingsphere-infra/shardingsphere-infra-context/src/main/java/org/apache/shardingsphere/infra/context/metadata/MetaDataContextsBuilder.java (hunk 1 above)
Comment (strongduanmu) [reply shardingsphere-infra/shardingsphere-infra-context/src/main/java/org/apache/shardingsphere/infra/context/metadata/MetaDataContextsBuilder.java:40]: > materials? or material?

@tristaZero `ShardingSphereRulesBuilderMaterials` is named with reference to `SchemaBuilderMaterials`. Should the style be consistent?

File: shardingsphere-infra/shardingsphere-infra-common/src/main/java/org/apache/shardingsphere/infra/config/properties/ConfigurationPropertyKey.java (hunk 2 above)
Comment (strongduanmu) [reply shardingsphere-infra/shardingsphere-infra-common/src/main/java/org/apache/shardingsphere/infra/config/properties/ConfigurationPropertyKey.java:140]: > how about 'allow tables with the same name exist '? Or something like that.

@tristaZero `allow-duplicate-single-table`?

Comment (strongduanmu) [reply shardingsphere-infra/shardingsphere-infra-common/src/main/java/org/apache/shardingsphere/infra/config/properties/ConfigurationPropertyKey.java:140]: @tristaZero Changing to `allow...` will cause the default value to be `true`, and when user checks the single table, it needs to be set to `false`, which is the opposite of using other parameters. Is it better to change to `check-single-table-duplicate-enabled`?

**Review by tristaZero — APPROVED** [review]

**Pull request merged**